control/endpoints.go

```diff
@@ -11,70 +11,66 @@ import (
+// HTTPServer API boundary
 type Endpoints struct {
 	bus *events.EventBus
 }
 
 // PostHandler is an adapter which allows a normal function to serve itself and
-// handle incoming HTTP POST requests. Also allows us to pass through App state
-// in a more organized fashion.
-type PostHandler func(http.ResponseWriter, *http.Request)
-
-// ServeHTTP implements intermediate endpoint behavior for POST
-// requests. Subsequently calls actual handler implementations defined below.
-func (ph PostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	log.Debugf("control: '%s %s' requested", r.Method, r.URL)
+// handle incoming HTTP POST requests, and allows us to pass thru EventBus to
+// handlers
+type PostHandler func(*http.Request) (interface{}, int)
 
+func (pw PostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if r.Method != http.MethodPost {
 		failedStatus := http.StatusNotImplemented
-		log.Errorf("%s requires POST, not %s", r.URL, r.Method)
 		http.Error(w, http.StatusText(failedStatus), failedStatus)
 		return
 	}
-
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-
-	ph(w, r)
-
-	io.WriteString(w, "\n")
+	resp, status := pw(r)
+	switch status {
+	case http.StatusOK:
+		if resp != nil {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			json.NewEncoder(w).Encode(resp)
+		} else {
+			w.WriteHeader(http.StatusOK)
+			io.WriteString(w, "\n")
+		}
+	default:
+		http.Error(w, http.StatusText(status), status)
+	}
 }
 
 // PutEnviron handles incoming HTTP POST requests containing JSON environment
 // variables and updates the environment of our current ContainerPilot
-// process. Returns null HTTP response.
-func (e Endpoints) PutEnviron(w http.ResponseWriter, r *http.Request) {
+// process. Returns empty response or HTTP422.
+func (e Endpoints) PutEnviron(r *http.Request) (interface{}, int) {
 	var postEnv map[string]string
-
-	errFunc := func(err error) {
-		failedStatus := http.StatusUnprocessableEntity
-		log.Errorf("'%v %v' request unprocessable due to error:\n%v", r.Method, r.URL, err)
-		http.Error(w, http.StatusText(failedStatus), failedStatus)
-	}
-
 	jsonBlob, err := ioutil.ReadAll(r.Body)
 	defer r.Body.Close()
 	if err != nil {
-		errFunc(err)
+		return nil, http.StatusUnprocessableEntity
 	}
-
 	err = json.Unmarshal(jsonBlob, &postEnv)
 	if err != nil {
-		errFunc(err)
+		return nil, http.StatusUnprocessableEntity
 	}
-
 	for envKey, envValue := range postEnv {
 		os.Setenv(envKey, envValue)
 	}
+	return nil, http.StatusOK
 }
 
 // PostReload handles incoming HTTP POST requests and reloads our current
-// ContainerPilot process configuration. Returns null HTTP response.
-func (e Endpoints) PostReload(w http.ResponseWriter, r *http.Request) {
+// ContainerPilot process configuration.  Returns empty response or HTTP422.
+func (e Endpoints) PostReload(r *http.Request) (interface{}, int) {
```

Comment:
Not yet, because I haven't written any of the endpoints that return anything other than nil. We can probably tighten up the interface at that point.